Fix spawn declaration on Windows

piotr-cz opened the pull request:
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

When configured a service like 'appium' in configuration, @wdio/appium-service launcher tries to spawn `PROJECT_DIR\appium\build\lib\main.js` file.

On Windows OS this results in error:
```
ERROR @wdio/cli:utils: A service failed in the 'onPrepare' hook
Error: spawn UNKNOWN
    at ChildProcess.spawn (internal/child_process.js:371:11)
    at spawn (child_process.js:561:9)
    at _startAppium (PROJECT_DIR\node_modules\@wdio\appium-service\build\launcher.js:54:44)
```

This happens due to fact that shebangs (`#!/usr/bin/env node`) don't work on Windows and command should be executed explicitly via node:
```js
spawn('node', ['PROJECT_DIR\node_modules\@wdio\appium-service\build\launcher.js', arg1, arg2])
```

instead of
```js
spawn('PROJECT_DIR\node_modules\@wdio\appium-service\build\launcher.js', [arg1, arg2])
```

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/technical-committee

~~~
Workaround:

in wdio.config.js use [Options](https://www.npmjs.com/package/@wdio/appium-service#options)
```js
appium: {
  command: 'node',
  args: [require.resolve('appium')]
}
```

christian-bromann commented on packages/wdio-appium-service/src/launcher.js:
```diff
@@ -39,7 +39,7 @@ export class AppiumLauncher {
     _startAppium(command, args, waitStartTime, callback) {
         log.debug(`Will spawn Appium process: ${command} ${args.join(' ')}`)
-        let process = spawn(command, args, { stdio: ['ignore', 'pipe', 'pipe'] })
+        let process = spawn('node', [command, ...args], { stdio: ['ignore', 'pipe', 'pipe'] })
```
What happens if Appium is installed globally? This wouldn't work then anymore.

piotr-cz replied:
AFAIK command is an absolute path to appium.
I just replaced it by PROJECT_DIR placeholder in example

christian-bromann replied:
as you can see in 
```js
this.command = appiumConfig.command || this._getAppiumCommand()
```
the command can be either an absolute path to the launcher (as in your example) or just something like `appium` in case it is installed globally on the machine. For latter your patch would break.

piotr-cz replied:
Path to appium executable is retrieved via [require.resolve](https://github.com/webdriverio/webdriverio/blob/6adfa2166f0656d4d61eb20c781b490374687e64/packages/wdio-appium-service/src/launcher.js#L79).

The point is that sole file name is not enough to spawn it on Windows.

piotr-cz replied:
This should work when Appium is installed globally.
Unless somebody found a workaround for this problem by and used `appium: { command: 'node appium' }` in wdio.conf.js config

piotr-cz replied:
Update: you are right. It's not possible to run global packages via node.

piotr-cz replied:
Or even shorter but mixing concepts of command and args together
```js
async onPrepare(config) {
  if (appiumConfig.command) {
    this.commmand = appiumConfig.command
  } else {
    this.command = 'node'
    this.appiumArgs = this._getAppiumCommand()
  }

  this.appiumArgs.push(...this._cliArgsFromKeyValue(appiumConfig.args || {}))
}
```

piotr-cz replied:
This should be resolved in https://github.com/webdriverio/webdriverio/pull/4136/commits/a49d7c8edc404bd1f5c4434136a023d54baae4b0

piotr-cz commented:
Note to myself:
Travis build failed because this PR breaks _Appium launcher › onPrepare › should start [Appium test](https://github.com/piotr-cz/webdriverio/blob/patch-1/packages/wdio-appium-service/tests/launcher.test.js#L92-L102)_ which inspects first argument for the spawn method against module path - but now receives `node`

mgrybyk commented:
Why don't you just pass desired command and arguments instead?

piotr-cz commented:
@mgrybyk because it doesn't work and that's why I started this PR.

On Windows OS there is no concept of executable permission on files or shebangs.
You can spawn only executables (`node.exe`) and pass script as an argument (`launcher.js`).

This is the isolated problem:
```sh
node -e "require('child_process').spawn('node_modules/@wdio/appium-service/build/launcher.js')

internal/child_process.js:371
    throw errnoException(err, 'spawn');
    ^

Error: spawn UNKNOWN
    at ChildProcess.spawn (internal/child_process.js:371:11)
    at Object.spawn (child_process.js:561:9)
    at [eval]:1:26
    at Script.runInThisContext (vm.js:124:20)
    at Object.runInThisContext (vm.js:314:38)
    at Object.<anonymous> ([eval]-wrapper:9:26)
    at Module._compile (internal/modules/cjs/loader.js:816:30)
    at evalScript (internal/process/execution.js:60:25)
    at internal/main/eval_string.js:16:1
```

However this one works:
```sh
node -e "require('child_process').spawn('node', ['node_modules/@wdio/appium-service/build/launcher.js'])
```

mgrybyk commented:
Have you tried passing node as command and 'node_modules/@wdio/appium-service/build/launcher.js' as argument?

It should work without any additional changes. 
@piotr-cz

christian-bromann commented on packages/wdio-appium-service/src/launcher.js:
```diff
@@ -19,8 +19,17 @@ export class AppiumLauncher {
+            this.commmand = appiumConfig.command
+        } else {
+            this.command = 'node'
+            this.appiumArgs = this._getAppiumCommand()
```
`_getAppiumCommand` returns a string , it should be:
```suggestion
            this.appiumArgs.push(this._getAppiumCommand())
```

christian-bromann replied:
Also can you add a comment mentioning the windows situation?

piotr-cz replied:
Done.

piotr-cz commented:
I implemented second option in https://github.com/webdriverio/webdriverio/pull/4136/commits/a49d7c8edc404bd1f5c4434136a023d54baae4b0.

Also adjusted tests for wdio-appium-service package in https://github.com/webdriverio/webdriverio/pull/4136/commits/9d2980b948e48b76ef3df436ff19d81243fb7b10.

However I didn't test manually on Linux or MacOS

piotr-cz commented:
What do you think about renaming __getAppiumCommand_ method to __getAppiumModulePath_?

mgrybyk commented:
works fine for me!

christian-bromann commented:
Released in `v5.10.10`

piotr-cz commented:
Thank you guys for help
